File: lib/scripts/nc.py
#!/usr/bin/env python

import os
import socket
import sys
import threading


class Sender(threading.Thread):
    def __init__(self, sock):
        super(Sender, self).__init__()
        self.sock = sock

    def run(self):
        line = sys.stdin.readline()
        while line:
            try:
                sock.sendall(line)
            except Exception as e:
                try:
                    sock.shutdown(socket.SHUT_RDWR)
                    sock.close()
                except Exception:
                    pass
                exit(1)
            line = sys.stdin.readline()


try:
    sock = socket.create_connection(("localhost", sys.argv[1]), timeout=None)
except Exception as e:
    exit(1)

sender = Sender(sock)
sender.daemon = True
sender.start()

# HACK: Try to read from the socket, which will block as long as the connection is OK since
# the agent isn't sending anything on the connection. As soon as this returns, it means the
# connection was closed on the other end (agent times out the connection if it does not receive
# logs during 1 minute). This is to prevent losing logs because after such a disconnect, the first
# message sent on the socket will succeed, even though it cannot possibly have been received.
try:
    sock.recv(1)
    sock.shutdown(socket.SHUT_RDWR)
    sock.close()
except Exception:
    pass
exit(1)

File: lib/scripts/nc.rb
#!/usr/bin/env ruby

require 'socket'

class Sender < Thread
  def initialize(sock)
    super()
    @sock = sock
  end

  def run
    line = $stdin.readline
    while line
      begin
        @sock.send(line, 0)
      rescue Exception => e
        begin
          @sock.shutdown(Socket::SHUT_RDWR)
          @sock.close
        rescue Exception
        end
        exit(1)
      end
      line = $stdin.readline
    end
  end
end

begin
  sock = TCPSocket.new("localhost", ARGV[0])
rescue Exception => e
  exit(1)
end

sender = Sender.new(sock)
sender.daemon = true
sender.start

# HACK: Try to read from the socket, which will block as long as the connection is OK since
# the agent isn't sending anything on the connection. As soon as this returns, it means the
# connection was closed on the other end (agent times out the connection if it does not receive
# logs during 1 minute). This is to prevent losing logs because after such a disconnect, the first
# message sent on the socket will succeed, even though it cannot possibly have been received.
begin
  sock.recv(1)
  sock.shutdown(Socket::SHUT_RDWR)
  sock.close
rescue Exception
end
exit(1)
